pkg/logs/input/journald/tailer_systemd.go

```diff
@@ -0,0 +1,238 @@
+	}()
+	for {
+		select {
+		case <-t.stop:
+			// stop tailing journal
+			return
+		default:
+			n, err := t.journal.Next()
+			if err != nil && err != io.EOF {
+				err := fmt.Errorf("cant't tail journal %s: %s", t.journalPath(), err)
+				t.source.Status.Error(err)
+				log.Error(err)
+				return
+			}
+			if n < 1 {
+				// no new entry
+				t.journal.Wait(defaultWaitDuration)
+				continue
+			}
+			entry, err := t.journal.GetEntry()
+			if err != nil {
+				log.Warnf("Could not retrieve journal entry: %s", err)
+				continue
+			}
+			if t.shouldDrop(entry) {
+				continue
+			}
+			t.outputChan <- t.toMessage(entry)
+		}
+	}
+}
+
+// shouldDrop returns true if the entry should be dropped,
+// returns false otherwise.
+func (t *Tailer) shouldDrop(entry *sdjournal.JournalEntry) bool {
+	unit, exists := entry.Fields[sdjournal.SD_JOURNAL_FIELD_SYSTEMD_UNIT]
+	if !exists {
+		return false
+	}
+	if _, blacklisted := t.blacklist[unit]; blacklisted {
+		// drop the entry
+		return true
+	}
+	return false
+}
+
+// toMessage transforms a journal entry into a message.
+// A journal entry has different fields that may vary depending on its nature,
+// for more information, see https://www.freedesktop.org/software/systemd/man/systemd.journal-fields.html.
+func (t *Tailer) toMessage(entry *sdjournal.JournalEntry) message.Message {
+	return message.New(
+		t.getContent(entry),
+		t.getOrigin(entry),
+		t.getStatus(entry),
+	)
+}
+
+// getContent returns all the fields of the entry as a json-string,
+// remapping "MESSAGE" into "message" and bundling all the other keys in a "journald" attribute.
+// ex:
+// * journal-entry:
+//  {
+//    "MESSAGE": "foo",
+//    "_SYSTEMD_UNIT": "foo",
+//    ...
+//  }
+// * message-content:
+//  {
+//    "message": "foo",
+//    "journald": {
+//      "_SYSTEMD_UNIT": "foo",
+//      ...
+//    }
+//  }
+func (t *Tailer) getContent(entry *sdjournal.JournalEntry) []byte {
+	payload := make(map[string]interface{})
+	fields := entry.Fields
+	if message, exists := fields[sdjournal.SD_JOURNAL_FIELD_MESSAGE]; exists {
+		payload["message"] = message
+		delete(fields, sdjournal.SD_JOURNAL_FIELD_MESSAGE)
+	}
+	payload["journald"] = fields
+
+	content, err := json.Marshal(payload)
+	if err != nil {
+		// ensure the message has some content if the json encoding failed
+		value, _ := entry.Fields[sdjournal.SD_JOURNAL_FIELD_MESSAGE]
+		content = []byte(value)
+	}
+
+	return content
+}
+
+// getOrigin returns the message origin computed from the journal entry
+func (t *Tailer) getOrigin(entry *sdjournal.JournalEntry) *message.Origin {
+	origin := message.NewOrigin(t.source)
+	origin.Identifier = t.Identifier()
+	origin.Offset, _ = t.journal.GetCursor()
+	// set the service and the source attributes of the message,
+	// those values are still overridden by the integration config when defined
+	applicationName := t.getApplicationName(entry)
+	origin.SetSource(applicationName)
+	origin.SetService(applicationName)
```

Comment:
Why do we need **both** source and service to be set to the same value ?

Reply:
It's a special requirement from @NBParis, we need at least the source to be set to the application name to ensure integrations are installed correctly.